clippy_lints/src/matches.rs

```diff
@@ -245,7 +245,74 @@ declare_clippy_lint! {
     "a wildcard pattern used with others patterns in same match arm"
 }
 
-declare_lint_pass!(Matches => [
+declare_clippy_lint! {
+    /// **What it does:** Checks for matches being used to destructure a single-variant enum
+    /// or tuple struct where a `let` will suffice.
+    ///
+    /// **Why is this bad?** Just readability – `let` doesn't nest, whereas a `match` does.
+    ///
+    /// **Known problems:** None.
+    ///
+    /// **Example:**
+    /// ```rust
+    /// enum Wrapper {
+    ///     Data(i32),
+    /// }
+    ///
+    /// let wrapper = Wrapper::Data(42);
+    ///
+    /// let data = match wrapper {
+    ///     Wrapper::Data(i) => i,
+    /// };
+    /// ```
+    ///
+    /// The correct use would be:
+    /// ```rust
+    /// enum Wrapper {
+    ///     Data(i32),
+    /// }
+    ///
+    /// let wrapper = Wrapper::Data(42);
+    /// let Wrapper::Data(data) = wrapper;
+    /// ```
+    pub INFALLIBLE_DESTRUCTURING_MATCH,
+    style,
+    "a `match` statement with a single infallible arm instead of a `let`"
+}
+
+declare_clippy_lint! {
+    /// **What it does:** Checks for useless match that binds to only one value.
+    ///
+    /// **Why is this bad?** Readability and needless complexity.
+    ///
+    /// **Known problems:** None.
```

Comment:
Will there be false-positives related to temporary values?

```rust
// I couldn't come up with a realistic example...

match io::stdout().lock() {
    mut stdout => writeln!(stdout, "hi").unwrap(), // ok
}

let mut stdout = io::stdout().lock(); // error[E0716]: temporary value dropped while borrowed
writeln!(stdout, "hi").unwrap();
```

Reply:
Yes it does. Any idea to prevent this? I don't think there is an easy way in Clippy, since this check is pretty deep inside of MIR in rustc.

Reply:
If it sounds out of scope of Clippy, maybe we should restrict applicability for non-wild binding.
I don't know if for wild binding this kind of situation could happen though.

Reply:
Yeah, I think we should use `MaybeIncorrect` for the `let` suggestion. The wildcard can stay `MachineApplicable`, since we don't bind temp values there.